src/Controls/src/SourceGen/RouteSourceGenerator.cs

```diff
@@ -0,0 +1,178 @@
+{
+	[Generator(LanguageNames.CSharp)]
+	public class RouteSourceGenerator : ISourceGenerator
+	{
+		static readonly string Indent = new string('\x20', 8);
+		const string AutoGeneratedHeaderText = """
+			//------------------------------------------------------------------------------
+			// <auto-generated>
+			//     This code was generated by a .NET MAUI source generator.
+			//
+			//     Changes to this file may cause incorrect behavior and will be lost if
+			//     the code is regenerated.
+			// </auto-generated>
+			//------------------------------------------------------------------------------
+			""";
+
+		public void Initialize(GeneratorInitializationContext context)
+		{
+			//#if DEBUG
+			//			if (!System.Diagnostics.Debugger.IsAttached)
+			//			{
+			//				System.Diagnostics.Debugger.Launch();
+			//			}
+			//#endif
+			context.RegisterForSyntaxNotifications(() => new RouteSyntaxReceiver());
+		}
+
+		public void Execute(GeneratorExecutionContext context)
+		{
+			if (context.SyntaxReceiver is not RouteSyntaxReceiver syntaxReceiver)
+			{
+				return;
+			}
+
+			var mauiApp = syntaxReceiver.MauiApp;
+			var mauiStartup = syntaxReceiver.MauiStartup;
+			var routedPages = syntaxReceiver.RoutedPages;
+			var shellPages = syntaxReceiver.ShellPages;
+
+			var routes = new StringBuilder();
+			var services = new StringBuilder();
+
+			GenerateCode(context, syntaxReceiver, ref routes, ref services, Singleton);
+			GenerateCode(context, syntaxReceiver, ref routes, ref services, Scoped);
+			GenerateCode(context, syntaxReceiver, ref routes, ref services, Transient);
+
+			string shellRoutes;
+
+			foreach (var shellPage in shellPages)
+			{
+				shellRoutes = $$"""
+				{{AutoGeneratedHeaderText}}
+
+				namespace {{GetNamespace(shellPage)}};
+
+				partial class {{shellPage.Identifier.ValueText}}
+				{
+				    static partial void RegisterRoutes();
+
+				    static partial void RegisterRoutes()
+				    {
+				        {{routes.ToString().Trim()}}
+				    }
+				}
+
+				""";
+
+				context.AddSource($"{shellPage.Identifier.ValueText}.sg.cs", shellRoutes);
+			}
+
+			if (mauiStartup is not null)
+			{
+				context.AddSource("MauiProgram.sg.cs", $$"""
+				{{AutoGeneratedHeaderText}}
+
+				using Microsoft.Extensions.DependencyInjection;
+				
+				namespace {{GetNamespace(mauiStartup)}};
+
+				static partial class MauiProgram
+				{
+				    static partial void ConfigureDependencies(this global::Microsoft.Maui.Hosting.MauiAppBuilder builder);
+
+				    static partial void ConfigureDependencies(this global::Microsoft.Maui.Hosting.MauiAppBuilder builder)
+				    {
+				        {{services.ToString().Trim()}}
+				    }
+				}
+
+				""");
+			}
+
+			static void GenerateCode(GeneratorExecutionContext context, RouteSyntaxReceiver syntaxReceiver, ref StringBuilder routes, ref StringBuilder services, string lifetime)
+			{
+				if (lifetime is null)
+				{
+					return;
+				}
+
+				var addComment = true;
+				SymbolInfo symbolInfo;
```

Comment:
nit: (and I'm not sure)
```suggestion
				SymbolInfo? symbolInfo;
```

Reply:
Since the return value is non-nullable from the Semantic Model of the SG, didn't consider it as marking as nullable.